src/main/java/edu/harvard/iq/dataverse/engine/command/impl/UpdateHarvestedDatasetCommand.java

```diff
@@ -0,0 +1,202 @@
+                existingDataset.getFiles().get(fileIndex).setChecksumType(newFileMetadata.getDataFile().getChecksumType());
+                existingDataset.getFiles().get(fileIndex).setChecksumValue(newFileMetadata.getDataFile().getChecksumValue());
+                
+                // Point the newly-harvested filemetadata to the existing datafile:
+                newFileMetadata.setDataFile(existingDataset.getFiles().get(fileIndex));
+                
+                // Make sure this new FileMetadata is the only one attached to this existing file:
+                existingDataset.getFiles().get(fileIndex).setFileMetadatas(new ArrayList<>(1));
+                existingDataset.getFiles().get(fileIndex).getFileMetadatas().add(newFileMetadata);
+                // (we don't want any cascade relationships left between this existing
+                // dataset and this version, since we are going to attemp to delete it).
+                
+                // Drop the file from the index map:
+                existingFilesIndex.remove(storageIdentifier);
+            }
+        }
+        
+        // @todo? check if there's anything special that needs to be done for things
+        // like file categories
+                
+        List<String> solrIdsOfDocumentsToDelete = new ArrayList<>();
+
+        // Go through the existing files and delete the ones that are 
+        // no longer present in the version that we have just harvesed:
+        for (FileMetadata oldFileMetadata : existingDataset.getVersions().get(0).getFileMetadatas()) {
+            DataFile oldDataFile = oldFileMetadata.getDataFile();
+            String storageIdentifier = oldDataFile.getStorageIdentifier();
+            // Is it still in the existing files map? - that means it is no longer
+            // present in the newly-harvested version
+            if (StringUtil.isEmpty(storageIdentifier) || existingFilesIndex.containsKey(storageIdentifier)) {
+                solrIdsOfDocumentsToDelete.add(solrDocIdentifierFile + oldDataFile.getId());
+                existingDataset.getFiles().remove(oldDataFile);
+                // Files from harvested datasets are removed unceremoniously, 
+                // directly in the database. No need to bother calling the 
+                // DeleteFileCommand on them. We'll just need to remember to purge
+                // them from Solr as well (right below)
+                ctxt.em().remove(ctxt.em().merge(oldDataFile));
+                // (no need to explicitly remove the oldFileMetadata; it will be 
+                // removed with the entire old version is deleted)
+            }
+        }
+                
+        // purge all the SOLR documents associated with the files
+        // we have just deleted:
+        if (!solrIdsOfDocumentsToDelete.isEmpty()) {
+            ctxt.index().deleteHarvestedDocuments(solrIdsOfDocumentsToDelete);
```

Comment:
I don't *think* you need to do this - the reindex will drop any docs that don't correspond to current files.

Reply:
I thought about it - but I figured I would leave the delete call in place. It's so easy to do there - I already know the ids of the documents that need to be purged; so, less work for the reindex code, and I don't think it can possibly hurt.